Examples/ByFeature/FunctionalInjection/ShoppingSteps.fs

```diff
@@ -1,17 +1,36 @@
 ﻿module ShoppingSteps
 
+open System
 open NUnit.Framework
 open TickSpec
 
-type Wallet = WalletDollars of int
+type CatalogItem = {Id: int; Description: string; Price: decimal}
+type Year = Year of int
+type Catalog = Catalog of Year:Year * Items:(CatalogItem [])
+type Catalogs = Map<Year, Catalog>
+type Receipt = Receipt of Date:DateTime * Lines:(string [])
 
-let [<Given>] ``I have (.*) dollars in my wallet`` (amount:int) =
-    WalletDollars amount // The output of function gets stored by type
+let [<Given>] ``no catalogs prior to 2018``() =
+    Map.empty: Catalogs
 
-// Extra parameters are restored from store by its type, in this case it is the wallet
-let [<When>] ``I buy an item for (.*) dollars`` cost (WalletDollars wallet) =
-    // Wallet value is replaced in store thanks to this output
-    WalletDollars (wallet - cost)
+// Table parameters can be bound to record type arrays
+let [<Given>] ``the (.+) product catalog:`` (yearInt: int) (items: CatalogItem[]) (catalogs: Catalogs) =
+    let year = Year yearInt
+    let catalog = Catalog(year, items)
+    Map.add year catalog catalogs
 
-let [<Then>] ``My wallet contains (.*) dollars`` (amount:int) (WalletDollars wallet) =
-    Assert.AreEqual(amount, wallet)
\ No newline at end of file
+// Bullet list is bound to string[]
+let [<When>] ``I make a purchase on (.*):`` (purchaseDate: DateTime) (orderItems: string[]) (catalogs: Catalogs) =
+    let Catalog(Items=catalogItems) = catalogs |> Map.find (Year purchaseDate.Year)
+    let receiptLines =
+        orderItems
+        |> Array.map (fun x -> catalogItems |> Array.find (fun y -> y.Description = x))
+        |> Array.map (fun x -> sprintf "%s: $%.2f" x.Description x.Price)
+    Receipt(purchaseDate, receiptLines)
+
+// DocString is passed as a regular string following captures
+let [<Then>] ``the receipt dated (.+) includes:`` (expectedDate: DateTime) (expected: string) (receipt: Receipt) =
+    let expectedLines = expected.Split('\n') |> Array.map(fun s -> s.Trim())
+    let Receipt(Date=receiptDate; Lines=receiptLines) = receipt
+    Assert.AreEqual(expectedDate, receiptDate)
```

Comment:
While Assert.AreEqual is ironically the most approachable, I favor having unquote expressions (`test <@ expected = actual @>`) in order to have consistent assertions independent of test framework (in this case it makes little difference, but the fail messages can be better too; also consider combining both into a

    test <@ expr1
                  && expr2 @>)